sky/execution.py

```diff
@@ -378,55 +379,120 @@ def spot_launch(
+        if os.path.isfile(os.path.abspath(os.path.expanduser(src))):
+            copy_mounts_with_file_in_src[dst] = src
+            continue
+        bucket_name = spot.constants.SPOT_FM_BUCKET_NAME.format(
             username=getpass.getuser(),
-            hash=f'{run_id}-{i}',
+            id=f'{run_id}-{i}',
         )
-        new_storage_mounts[dst] = storage.Storage.from_yaml_config({
+        new_storage_mounts[dst] = storage_lib.Storage.from_yaml_config({
             'name': bucket_name,
             'source': src,
             'persistent': False,
             'mode': 'COPY',
         })
         logger.info(
-            f'Translated local file mount {dst} to cloud storage {bucket_name}.'
-        )
+            f'Folder in local file mount {src!r} will be synced to SkyPilot '
+            f'storage {bucket_name}.')
+
+    # Step 3: Translate local file mounts with file in src to SkyPilot storage.
+    # Hard link the files in src to a temporary directory, and upload folder.
+    original_storage_mounts = {}
+    if task.storage_mounts is not None:
+        original_storage_mounts = task.storage_mounts
+    local_fm_path = os.path.join(
+        tempfile.gettempdir(),
+        spot.constants.SPOT_FM_LOCAL_TMP_DIR.format(id=run_id))
+    os.makedirs(local_fm_path, exist_ok=True)
+    file_bucket_name = spot.constants.SPOT_FM_FILE_ONLY_BUCKET_NAME.format(
+        username=getpass.getuser(), id=run_id)
+    if copy_mounts_with_file_in_src:
+        src_to_file_id = dict()
+        for i, src in enumerate(set(copy_mounts_with_file_in_src.values())):
+            src_to_file_id[src] = i
+            os.link(os.path.abspath(os.path.expanduser(src)),
+                    os.path.join(local_fm_path, f'file-{i}'))
+
+        new_storage_mounts[
+            spot.constants.
+            SPOT_FM_REMOTE_TMP_DIR] = storage_lib.Storage.from_yaml_config({
+                'name': file_bucket_name,
+                'source': local_fm_path,
+                'persistent': False,
+                'mode': 'MOUNT',
+            })
+        if spot.constants.SPOT_FM_REMOTE_TMP_DIR in original_storage_mounts:
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    'Failed to translate file mounts, due to the default '
+                    f'destination {spot.constants.SPOT_FM_REMOTE_TMP_DIR} '
+                    'is taken.')
+        sources = list(src_to_file_id.keys())
+        sources_str = '\n\t'.join(sources)
+        logger.info(
+            f'Sources with file will be synced to cloud storage {bucket_name}:'
+            f'\n\t{sources_str}')
     task.update_storage_mounts(new_storage_mounts)
 
+    # Step 4: Upload storage from sources
     # Copy the local source to a bucket. The task will not be executed locally,
     # so we need to copy the files to the bucket manually here before sending to
     # the remote spot controller.
```

Comment:
Move L475-477?

Reply:
I was confused by original L467-478 here
```
    # Copy the local source to a bucket. The task will not be executed locally,
    # so we need to copy the files to the bucket manually here before sending to
    # the remote spot controller.
```
on second look, they do describe `task.sync_storage_mounts()` line below?  If so, I think we should change "copy" to "upload"/"sync" to avoid confusion.

Reply:
Ahh, good point! Just replace the `copy` with `upload`. Thanks!